### ISSUE TITLE
One Click Heroku Deploy

### ISSUE THREAD

**Issue by tomkysar** [AugurProject/augur-node#292]
Augur Node should be deployable with Heroku's one-click deploy.

Please join the #dev or #bounties channel of our [Discord](http://augur.net/invite) with any questions. 

Bounty posted with [Gitcoin](https://gitcoin.co).

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 300.0 DAI (300.0 USD @ $1.0/DAI) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/AugurProject/augur-node/292/331).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $6,378.38 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__Work has been started on the 300.0 DAI (300.0 USD @ $1.0/DAI) funding by__: 
 1. [@JonnyBurger](https://gitcoin.co/profile/JonnyBurger)  

 On the above list? Please leave a comment to let the funder (@tomkysar) know your  plan. If you don't leave a comment, the funder may expire your submission at their discretion. 

 * Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/AugurProject/augur-node/292/331)
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $7,174.9 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by JonnyBurger** [comment]
Hello there! Just trying out Gitcoin.

This is what I did:

* Added `ENDPOINT_HTTP` and `ENDPOINT_WS` Infura env variables as default to the Heroku `app.json`.
 ---> I find that setting these variables does not seem to actually override the Ethereum node parameters as it says in the README and nowhere in the app is checked for any environment variables. So I have added that functionality.
* Ignores `package-lock.json` for Heroku. Having both `yarn.lock` and `package-lock.json` will negate some of the benefits of lockfiles and will cause Heroku to fail. See [here](https://help.heroku.com/0KU2EM53/why-is-my-node-js-build-failing-because-of-conflicting-lock-files). For now, I ignored one of the files in `.slugignore`, but this should be fixed by you
* Initialize a git repo during the Heroku build process. You have `shared-git-hooks` enabled during the build process, but it fails when there is no Git repo such as during the Heroku build process. This is why I added a `preinstall` script that runs `git init` if and only if the `.git` directory does not exist.
* Run migration and seeds before starting the server (added in `Procfile`)
* Heroku will not run unless you bind the server to `$PORT`. So I added that the HTTP (not HTTPS) server looks for a $PORT environment variable first and otherwise defaults. Heroku will proxy HTTP to HTTPS then using Let's Encrypt.

I also tried to make it run with Postgres using the Heroku Postgres addon.
However `knex` does seem to fail with any PG database. This is the stack trace:

<details>

```
Knex:warning - migration file "20180326123259_crowdsourcer_redeemed.js" failed
Knex:warning - migration failed with error: CREATE TABLE markets (
      "marketId" varchar(66) PRIMARY KEY NOT NULL,
      universe varchar(66) NOT NULL,
      "marketType" varchar(11) NOT NULL CONSTRAINT "enumMarketTypes" CHECK ("marketType" = 'binary' OR "marketType" = 'categorical' OR "marketType" = 'scalar'),
      "numOutcomes" integer NOT NULL CONSTRAINT "positiveNumOutcomes" CHECK ("numOutcomes" > 0),
      "minPrice" VARCHAR(255) NOT NULL,
      "maxPrice" VARCHAR(255) NOT NULL,
      "marketCreator" varchar(66) NOT NULL,
      "creationBlockNumber" integer NOT NULL CONSTRAINT "positiveMarketCreationBlockNumber" CHECK ("creationBlockNumber" > 0),
      "creationFee" VARCHAR(255) NOT NULL,
      "reportingFeeRate" VARCHAR(255) NOT NULL,
      "marketCreatorFeeRate" VARCHAR(255) NOT NULL,
      "marketCreatorFeesClaimed" VARCHAR(255) DEFAULT "0",
      "marketCreatorFeesCollected" VARCHAR(255) DEFAULT "0",
      "initialReportSize" VARCHAR(255) DEFAULT "0",
      category varchar(255) NOT NULL,
      tag1 varchar(255),
      tag2 varchar(255),
      volume VARCHAR(255) NOT NULL,
      "sharesOutstanding" VARCHAR(255) NOT NULL,
      "feeWindow" varchar(66),
      "endTime" integer NOT NULL CONSTRAINT "positiveEndTime" CHECK ("endTime" > 0),
      "finalizationTime" integer,
      "marketStateId" integer,
      "shortDescription" text NOT NULL,
      "longDescription" text,
      "scalarDenomination" text,
      "designatedReporter" varchar(66) NOT NULL,
      "designatedReportStake" VARCHAR(255),
      "resolutionSource" text,
      "numTicks" VARCHAR(255) NOT NULL,
      "consensusPayoutId" integer,
      "disputeRounds" integer,
      "isInvalid" boolean
    ) - column "0" does not exist
```

</details>
<br/>

Otherwise it would work when you just add the Heroku Postgres plugin in `app.json`.
So now it uses SQLite3 which I figured is fine because your Docker deploy method uses it also. 

You need to change the Heroku deploy button URL in `app.json`, I left my fork in there to demonstrate that it works: https://github.com/JonnyBurger/augur-node

My PR: https://github.com/AugurProject/augur-node/issues/292

My instance: https://agile-hollows-94803.herokuapp.com/
<img width="1317" alt="bildschirmfoto 2018-04-30 um 12 06 31" src="https://user-images.githubusercontent.com/1629785/39422743-f7437578-4c6e-11e8-84d4-dd420d53a365.png">

Cheers!

**Comment by vs77bb** [comment]
@tomkysar Have you been able to review @JonnyBurger's PR here yet? Hope all is well!

**Comment by weberV** [comment]
Issue Status: 1. Open 2. Started 3. **Submitted** 4. Done 

<hr>

__This issue now has a funding of 300.0 DAI (300.0 USD @ $1.0/DAI) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](http://localhost:8000/issue/AugurProject/augur-node/292/331).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $7,827.96 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by JonnyBurger** [comment]
Hi,

could you please accept the work on Gitcoin?
It expires in one day.

I delivered the work fulfilling all requirements.
Fixing Postgres with Knex is outside the scope of this issue I think.
I will still make another PR once Postgres works to enable it on Heroku as well.

**Comment by owocki** [comment]
hey @tomkysar did you see Jonny's PR?

**Comment by tomkysar** [comment]
cc: @pgebheim @epheph @tinybike 

I can re-bounty these issues - can someone review & give the next steps? Is the proposed solution sufficient?

**Comment by vs77bb** [comment]
Hi @tomkysar, I think we're still pending review / payout here if appropriate from the Augur side!

**Comment by JonnyBurger** [comment]
Hi there,

I did everything that I can do from my side to help you with this feature.  
This is taking far to long for you to review this. I think you should pay out the bounty.

cc: @owocki

**Comment by tinybike** [comment]
@JonnyBurger the main requirement of this bounty is that it must work with postgres / that's the whole point of the heroku deployment.

Sorry for the slow response.  The engineers assigned to review this didn't do so.

**Comment by JonnyBurger** [comment]
@tinybike It seems like you have your SQL structured in a way that is incompatible with Postgres.
This issue is much more fundamental than than Heroku deployment.

> the main requirement of this bounty is that it must work with postgres

The main requirement that I understood is to add the necessary files needed to deploy to Heroku. Postgres is not even mentioned in your task description.

**Comment by pgebheim** [comment]
The main point of the bounty was to have a working on click install for heroku. Since local sqlite databases are not a workable soltuion for a real heroku deploy, we either need Postgres compatibility or some other DBMS compatibility or it isn't a working one-click install.

**Comment by owocki** [comment]
hi all, kevin from gitcoin here!  we've been watching the convo internally here and wanted to provide the following suggestions:

first, its natural for some scope to be misunderstood or not fully laid out -- so we shouldn't feel bad about that.  software development is naturally iterative, and that's okay.

second, one of the things we do internally, when we use our product, is to plan for this type of iteration by offering a little more budget for tipping after the fact for extra scope.  

third, sometimes it just doesnt work out and we need to go our seperate ways.  in that case, i like to leave a tip for the bounty hunter for any value that was provided towards the end scope and re-bounty it.  

i dont know enough about the current scope to say whether the postgres stuff is do-able or is a heavy lift or not to know which option is the best for both parties.  but suffice to say we'd like to see you both work it out.

feel free to hit us up on slack with any questions; and enjoy your holiday everyone who's in the US 🇺🇸

**Comment by tomkysar** [comment]
@JonnyBurger, if you're willing to work out the Postgres stuff, we'd be more than happy to double the bounty. If not, I am also more than happy to close this as-is, and pay you out the full bounty amount currently. 

I sincerely apologize on behalf of the Foundation for the delay and confusion - Gitcoin has been great for us, and we want to see our bounty hunters succeed!

Thanks for chiming in as well @owocki, your feedback is very valuable.

**Comment by JonnyBurger** [comment]
Hi all,

for the record, Augur has reached out to me on Twitter and we have agreed that I will try to solve the Postgres issue as well and they are doubling the bounty for that.

Thanks everyone for trying to find a solution! Very much appreciated.

I have rebased the PR now and will work on the Postgres issue now.

**Comment by JonnyBurger** [comment]
Hi guys,

It worked, Augur now runs with Postgres as well! As you can see, it was not so trivial.

Here are some notes to help you understand the things I changed:

- **Behavior change:** The server starts listening before everything is synced. This is because Heroku will timeout if you don't bind to $PORT within 60 seconds of startup.
- Postgres does not support the datatype `unsigned integer`, so it just uses an integer with >= 0 constraint if it is not SQLite.
- `booleans` are `true/false` in Postgres but `0/1` in SQLite. This is why I added some casts throughout the code to normalize it.
- I increased the length of some fields like `orders.orderId` from 44 to 66, because otherwise the seed data won't fit. (In SQLite the specified string length had no effect, so it was no problem)
- Two constraints had the same name, which is not supported in Postgres. In Postgres it now has different names, but in SQLite it is the same so the schema is compatible.
- Had to double-quote some identifiers because they seem to be case sensitive. String literals only seem to work with single quotes on both databases.
- The join in the `all_participants` view was slightly reordered, otherwise an error would occur because some joins were performed before all data was selected.
- Changed `reporter` fields from a string to an integer in the seed data because it's an integer in the schema.
- Changed `isInvalid` from a integer to a boolean and standardized it throughout the code. This is backwards compatible because booleans are still just 0/1 in SQLite.
- Had to add some casts to `BigNumber`.
- Some queries were executed in parallel using the function from the `async` npm package. This caused some race conditions (for example, the transaction history was not correctly verified and some accounts ended up having negative account balances), so for Postgres, the queries are executed in sequential order instead. I made a switch so that for SQLite it stays parallel (however I believe this parallelism does not really bring performance because SQLite does not support parallel queries and the race conditions only started showing up because Postgres does support them, not entirely sure though).
- Upgraded some `int`s to `bigint`s. It's the same datatype in SQLite so this is also fully backwards compatible.

Also I implemented that you can run the Node on any network (Ropsten, Kovan, Rinkeby, etc.) and configure it using the Heroku UI :)

Please review carefully and let me know if it's all ok.

**Comment by pgebheim** [comment]
Hello!

Thank you so much for all this work. I'll review this as soon as possible. We're a little swamped with initial release happening on Monday, but I will get it reviewed shortly after that!

**Comment by JonnyBurger** [comment]
Hi there!
I have received the 300 DAI bounty through Gitcoin on August 2nd. Thanks a lot for that!
The other 300 DAI that were offered I haven't received yet. Also I saw that the PR is not yet merged!

Besides the bounty, I'm also interested in that you actually can use the Heroku button and Postgres integration.
Therefore, can you give an update on your side of this issue?

Cheers!

**Comment by petervecchiarelli** [comment]
@JonnyBurger Hey there,

Sorry again for the delay. I will send the remaining DAI reward within the next few minutes. WRT Postgres: in order to improve reliability we did a quick fix of using Sqlite files as snapshots. The logic isn't quite abstract enough to work for non Sqlite databases yet, but now that we are stable we can revisit.

**Comment by JonnyBurger** [comment]
Thanks for sending!
Sorry as well for letting it slip again. 

I don‘t quite understand the SQlite thing, but glad to hear you‘re on it. Let me know if I can help. Otherwise glad to have worked with you and hope to do it again!
